### ISSUE TITLE
Init fails because or wrong assumption

### ISSUE THREAD

**Issue by jackyhallyday** [revive-adserver/revive-adserver#960]
Hi,

I've just tried to install Revive Adserver and it failed because of the includes. For instance:

`require_once 'variables.php';`

It assumes that PHP will search in the current directory first, but that's not always the case.
I've changed the php's includes path so it searches in a special directory first, which contains a file called 'variables.php'. So the Revive Adserver script includes my variables.php instead of the expected one, and fails.

Fix: use absolute paths.
`require_once __DIR__ . '/variables.php';`

Cheers.

**Comment by andrewatfornax** [comment]
Hi @jackyhallyday,

I appreciate what you are saying, and I can see that it's a valid issue - but in 14 years of working on Revive Adserver (and its predecessors) this is the first time someone has raised this as an issue.

Given the effort required to change all the requires in the code base, and then test everything, it's not worth it. 

I think we can just take it as a requirement that Revive Adserver needs to not have a PHP configuration that makes it search somewhere else for files first. That can even be achieved just on the Revive Adserver host basis, so it's not something that will really stop you from doing that special searching for everything else on the server, if that's what you need.

Thanks.

**Comment by jackyhallyday** [comment]
> this is the first time someone has raised this as an issue
Maybe because people who can't even install Revive just move on?

> Given the effort required to change all the requires in the code base
It's super easy and quick. Almost every single require in the codebase begins with a relative path, __DIR__ or with the MAX_PATH constant. So I guess it was already taken care of, at some point, and somebody broke it.

The only problematic requires I found are located in the index.php:
`require_once 'init.php';`

And in the init.php:
```
require_once 'pre-check.php';
require_once 'init-parse.php';
require_once 'variables.php';
require_once 'constants.php';
```

By fixing them (with __DIR__ and a slash before the filename), Revive works just fine with my PHP config.

Well, not exactly, because **there's another bug**: if MySQL is configured with utf8mb4 for default collation, the database/tables creation fails.
To fix that, I had to create the database manually (with using utf8_bin for callation) before running the installation process.

**Comment by erikgeurts** [comment]
Hi Jack,
In your initial post you wrote:

> I've changed the php's includes path so it searches in a special directory first

As such, it might be that you've created a unique environment just for yourself, which turns out to be incompatible with our code. That doesn't necessarily mean that others will run into this very same issue (and based on our experience, as @andrewatfornax mentioned, we've actually never heard of anyone with this same issue).

Please consider opening a separate issue for the other problem you ran into.

**Comment by jackyhallyday** [comment]
> As such, it might be that you've created a unique environment just for yourself
Yes, but that's something PHP allows by changing the appropriate parameter in the php.ini.
I didn't change the PHP source files, if that's what you think. It's not a hack, it's a custom configuration.
http://php.net/manual/en/ini.core.php#ini.include-path

**Comment by erikgeurts** [comment]
I'm not suggesting that you've done something that's not "allowed". I'm just saying that whatever you've changed results in an environment that no longer performs the way the Revive Adserver code expects it to be, which is a setup without that custom configuration you created for yourself.
The fix you found seems to work for you, and there's nothing to stop you from applying it. But for most typical users, it should never be necessary.

**Comment by jackyhallyday** [comment]
I finally switched to Revive Adserver (from DFP), and I'm super pleased.

I'm using the async script integration on a separate domain. The cleanest way I found to patch the annoyance described in my first message is pretty simple:
In /init.php, I added
`set_include_path('.');`
before the includes. So Revive Adserver has the include path it expects, and everything's fine.

Maybe it wouldn't solve anything if I was using the Local Mode Tag, because I could have some custom code after. In that case, the workaround would be easy: save the current includes path (provided by get_includes_path()), run the Revive integration code, then restore the includes path.
It could be done in the Bannercode, directly in the source code.

I don't know if there are various entry/exit points for the different pages, but if there are one single entry point and one single exit point, get/set_include_path is a very simple way to provide Revive the environment it expects without having to make any other change to the source code.
